Mathlib/RingTheory/LocalRing/Module.lean

```diff
@@ -0,0 +1,269 @@
+    Module.Free R M := by
+  let I := Module.Free.ChooseBasisIndex k (k ⊗[R] M)
+  -- Let `b : I → k ⊗ M` be a `k`-basis.
+  let b : Basis I k _ := Module.Free.chooseBasis k (k ⊗[R] M)
+  letI : IsNoetherian k (k ⊗[R] (I →₀ R)) :=
+    isNoetherian_of_isNoetherianRing_of_finite k (k ⊗[R] (I →₀ R))
+  choose f hf using TensorProduct.mk_surjective R M k Ideal.Quotient.mk_surjective
+  -- By choosing an abitrary lift of `b` to `I → M`, we get a surjection `i : Rᴵ → M`.
+  let i := Finsupp.total I M R (f ∘ b)
+  have hi : Surjective i := by
+    rw [← LinearMap.range_eq_top, Finsupp.range_total]
+    exact LocalRing.span_eq_top_of_tmul_eq_basis (R := R) (f := f ∘ b) b (fun _ ↦ hf _)
+  have : Module.Finite R (LinearMap.ker i) := by
+    constructor
+    exact (Submodule.fg_top _).mpr (Module.FinitePresentation.fg_ker i hi)
+  -- We claim that `i` is actually a bijection,
+  -- hence an isomorphism exhibing `M` as the free module `Rᴵ`
+  refine Module.Free.of_equiv (LinearEquiv.ofBijective i ⟨?_, hi⟩)
+  -- By Nakayama's lemma, it suffices to show that `k ⊗ ker(i) = 0`.
+  rw [← LinearMap.ker_eq_bot, ← Submodule.subsingleton_iff_eq_bot,
+    ← LocalRing.subsingleton_tensorProduct (R := R), subsingleton_iff_forall_eq 0]
+  have : Function.Surjective (i.baseChange k) := i.lTensor_surjective _ hi
+  -- By construction, `k ⊗ i : kᴵ → k ⊗ M` is bijective.
+  have hi' : Function.Bijective (i.baseChange k) := by
+    refine ⟨?_, this⟩
+    rw [← LinearMap.ker_eq_bot (M := k ⊗[R] (I →₀ R)) (f := i.baseChange k),
+      ← Submodule.finrank_eq_zero (R := k) (M := k ⊗[R] (I →₀ R)),
+      ← Nat.add_right_inj (n := FiniteDimensional.finrank k (LinearMap.range <| i.baseChange k)),
+      LinearMap.finrank_range_add_finrank_ker (V := k ⊗[R] (I →₀ R)),
+      LinearMap.range_eq_top.mpr this, finrank_top]
+    simp only [FiniteDimensional.finrank_tensorProduct, FiniteDimensional.finrank_self,
+      FiniteDimensional.finrank_finsupp_self, one_mul, add_zero]
+    rw [FiniteDimensional.finrank_eq_card_chooseBasisIndex]
+  -- On the other hand, `m ⊗ M → M` injective => `Tor₁(k, M) = 0` => `k ⊗ ker(i) → kᴵ` injective.
+  have := @lTensor_injective_of_exact_of_exact_of_rTensor_injective
+    (N₁ := LinearMap.ker i) (N₂ := I →₀ R) (N₃ := M)
+    (f₁ := (𝔪).subtype) (f₂ := Submodule.mkQ 𝔪) inferInstance inferInstance inferInstance
+    inferInstance inferInstance inferInstance
+  intro x
+  apply @this (LinearMap.ker i).subtype i (LinearMap.exact_subtype_mkQ 𝔪)
+    (Submodule.mkQ_surjective _) (LinearMap.exact_subtype_ker_map i) hi H
+    (Module.Flat.lTensor_preserves_injective_linearMap _ Subtype.val_injective)
+  apply hi'.injective
+  rw [LinearMap.baseChange_eq_ltensor]
+  erw [← LinearMap.comp_apply (i.lTensor k), ← LinearMap.lTensor_comp]
+  rw [(LinearMap.exact_subtype_ker_map i).linearMap_comp_eq_zero]
+  simp only [LinearMap.lTensor_zero, LinearMap.zero_apply, map_zero]
+
+theorem free_of_flat_of_localRing [Module.FinitePresentation R P] [Module.Flat R P] :
```

Comment:
I think this is true for any finite module, maybe we can add a TODO (the proof is much harder IIRC).

Reply:
We're actually not that far away. We already have the equational criterion of flatness and finite flat = free is a (relatively) easy application of that. Anyways I added a TODO.